bonus/hkd.md

```diff
@@ -0,0 +1,367 @@
+# HKD - Higher-Kinded Data
+
+This file is literate haskell. If you have
+[`markdown-unlit`](https://hackage.haskell.org/package/markdown-unlit)
+installed, you can extract the code by running the following command:
+
+```
+$ markdown-unlit -h mtl.md mtl.md mtl.hs
+```
+
+This will allow you to work through the exercises using GHCi.
+
+In this document, we will explore a pattern called "higher-kinded
+data", which is when a data structure is parameterised by a
+`Functor`. This is becoming increasingly common in libraries like
+[`waargonaut`](https://hackage.haskell.org/package/waargonaut) (which
+we used for JSON encoding/decoding),
+[`dependent-sum`](https://hackage.haskell.org/package/dependent-sum),
+and
+[`dependent-map`](https://hackage.haskell.org/package/dependent-map),
+so it is worth knowing about.
+
+
+<details>
+<summary>Extensions and Imports</summary>
+
+```haskell
+{-# OPTIONS_GHC -Wall -Wno-unused-imports -Wno-unused-matches #-}
+
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE NoStarIsType #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+import Data.Functor.Compose (Compose(..))
+import Data.Functor.Identity (Identity(..))
+import qualified Data.Semigroup as Semigroup
+import Data.Kind (Type)
+import Data.Word (Word16)
+```
+
+</details>
+
+
+## Generalising `Conf`
+
+Consider the configuration types from level 7:
+
+```haskell
+newtype Port = Port Word16
+
+data Conf = Conf
+  { port       :: Port
+  , dbFilePath :: FilePath
+  }
+
+data PartialConf = PartialConf
+  { pcPort       :: Maybe (Semigroup.Last Port)
+  , pcDBFilePath :: Maybe (Semigroup.Last FilePath)
+  }
+```
+
+There's repeated structure here. We can generalise over the both of
+these structures by introducing a `Functor`-kinded type parameter:
+
+```haskell
+data Config f = Config
+  { _port :: f Port
+  , _dbFilePath :: f FilePath
+  }
+
+-- It is sometimes useful to provide a convenience function that uses
+-- pure to add all the `f` wrappers.
+-- Exercise: implement it.
+config :: Applicative f => Port -> FilePath -> Config f
+config = error "config"
+
+-- The forall in the constraint uses -XQuantifiedConstraints, available since
+-- GHC 8.6.1 . You could explicitly list them out instead:
+-- instance (Show (f Port), Show (f FilePath)) => ...
+deriving instance (forall a . Eq (f a)) => Eq (Config f)
+deriving instance (forall a . Show (f a)) => Show (Config f)
```

Comment:
Add a bit to say why we need to write the instances this way, and also that we don't have to do much work because GHC does the boring bit for us.